stl/inc/ranges

```diff
@@ -6821,6 +6822,496 @@ namespace ranges {
+    concept _Zip_is_common = (sizeof...(_RangeTypes) == 1 && (common_range<_RangeTypes> && ...)) //
+                          || (!(bidirectional_range<_RangeTypes> && ...) && (common_range<_RangeTypes> && ...)) //
+                          || ((random_access_range<_RangeTypes> && ...) && (sized_range<_RangeTypes> && ...));
+
+    template <class _CallbackType>
+    constexpr auto _Tuple_transform_closure(_CallbackType& _Callback) noexcept {
+        return [&_Callback]<class... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+            return tuple<invoke_result_t<_CallbackType&, _ViewTupleTypes>...>(
+                _STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples))...);
+        };
+    }
+
+    template <class _CallbackType, class _TupleType>
+    constexpr auto _Tuple_transform(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(noexcept(
+        _STD apply(_Tuple_transform_closure(_Callback), _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+        return _STD apply(_Tuple_transform_closure(_Callback), _STD forward<_TupleType>(_Tuple));
+    }
+
+    template <class _CallbackType>
+    constexpr auto _Tuple_for_each_closure(_CallbackType& _Callback) noexcept {
+        return [&_Callback]<class... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+            ((void) (_STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples))), ...);
+        };
+    }
+
+    template <class _CallbackType, class _TupleType>
+    constexpr void _Tuple_for_each(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+        noexcept(_STD apply(_Tuple_for_each_closure(_Callback), _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+        _STD apply(_Tuple_for_each_closure(_Callback), _STD forward<_TupleType>(_Tuple));
+    }
+
+    template <bool _IsConst, class... _Views>
+    concept _All_random_access = (random_access_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+    template <bool _IsConst, class... _Views>
+    concept _All_bidirectional = (bidirectional_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+    template <bool _IsConst, class... _Views>
+    concept _All_forward = (forward_range<_Maybe_const<_IsConst, _Views>> && ...);
+
+    template <class _ResultType, class... _LHSTupleTypes, class... _RHSTupleTypes>
+        requires(sizeof...(_LHSTupleTypes) == sizeof...(_RHSTupleTypes))
+    _NODISCARD constexpr _ResultType _Zip_get_smallest_iterator_sentinel_distance(
+        const tuple<_LHSTupleTypes...>& _Lhs_tuple, const tuple<_RHSTupleTypes...>& _Rhs_tuple)
+        // clang-format off
+        noexcept((noexcept(static_cast<_ResultType>(_STD declval<_LHSTupleTypes>() - _STD declval<_RHSTupleTypes>()))
+            && ...));
+    // clang-format on
+
+    template <class... _LHSTupleTypes, class... _RHSTupleTypes>
+        requires(sizeof...(_LHSTupleTypes) == sizeof...(_RHSTupleTypes))
+    _NODISCARD constexpr bool _Zip_evaluate_iterator_sentinel_equality(

```

Comment:
Could this be renamed to `_Zip_iterator_sentinel_equal`, and merged with the definition?